protocols/streaming/src/handler.rs

```diff
@@ -0,0 +1,306 @@
+use futures::future;
+use libp2p_core::{
+    upgrade::NegotiationError, InboundUpgrade, OutboundUpgrade, UpgradeError, UpgradeInfo,
+};
+use libp2p_swarm::{
+    protocols_handler::InboundUpgradeSend, KeepAlive, NegotiatedSubstream, ProtocolsHandler,
+    ProtocolsHandlerEvent, ProtocolsHandlerUpgrErr, SubstreamProtocol,
+};
+use smallvec::SmallVec;
+use std::{
+    collections::VecDeque,
+    convert::Infallible,
+    marker::PhantomData,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    task::Poll,
+    time::Instant,
+};
+
+use crate::{
+    InboundStreamId, OutboundStreamId, StreamHandle, StreamingCodec, StreamingConfig,
+    EMPTY_QUEUE_SHRINK_THRESHOLD,
+};
+
+pub(crate) type RefCount = Arc<PhantomData<u8>>;
```

Comment:
Do I understand correctly that the task of the handler is not woken up, when a `StreamHandle` with a `RefCount` is dropped? If so, a handler, if not polled due to some other event, might stay in `KeepAlive::Yes` forever, even though all its streams are closed, correct?

In `libp2p-relay` I have solved this with the help of a `oneshot`. Would that be an option here as well?

https://github.com/libp2p/rust-libp2p/blob/98bc5e64869d0d188384a91ba9287eff4eb27b0e/protocols/relay/src/handler.rs#L110-L119